torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -547,40 +543,41 @@ def _setup_flat_buffers(self) -> None:
         parameter trainability is changed (frozen or unfrozen) and
         ``_update_trainable`` is called.
         """
-
-        for device, per_rank_params in self._per_device_params.items():
-            # Only wipe the existing buckets if there are none
-            # (could be that this is called twice, when trainability changes)
-            if device not in self.buckets.keys():
-                self.buckets[device] = []
-
-            # Make parameters a view of the bucket
-            for dst_rank, params in enumerate(per_rank_params):
-                if len(params) > 0:
-
-                    # Clone the non-trainable params, if in a bucket it will get destroyed
-                    for param in filter(lambda x: not x.requires_grad, params):
+        for rank, param_groups in enumerate(self.partition_parameters()):
+            # Clone the non-trainable params, find the buffer size and dtype
+            # for the trainable params' bucket, and compile a list of the
+            # trainable params
+            buffer_size = 0
+            dtype = None
+            trainable_params = []
+            for param_group in param_groups:
+                for param in param_group["params"]:
+                    if (not _is_trainable(param)):
                         param.data = param.data.detach().clone()
-
-                    # Merge all the trainable params in a single bucket
-                    trainable_params = list(filter(_is_trainable, params))
-                    buffer_size = sum(map(lambda x: x.numel(), trainable_params))
-                    bucket = torch.empty(buffer_size, dtype=params[0].dtype, device=device)
-                    offset = 0
-
-                    for param in trainable_params:
-                        offset_next = offset + param.numel()
-                        bucket[offset:offset_next].copy_(param.data.flatten())
-                        param.data = bucket[offset:offset_next].view_as(param.data)
-                        offset = offset_next
-
-                    # Either replace the existing bucket, or create it
-                    if len(self.buckets[device]) == dst_rank:
-                        self.buckets[device].append(bucket)
                     else:
-                        self.buckets[device][dst_rank] = bucket
-                else:
-                    self.buckets[device].append(torch.zeros(1, device=device))
+                        buffer_size += param.numel()
+                        trainable_params.append(param)
+                    dtype = param.dtype
```

Comment:
For PyTorch optimizers, can we assume that the `dtype` of the parameters are always the same? This assumption was made in the old version, but I wanted to check.

Reply:
I don't think we can make this assumption. E.g., DDP bucket assignment does allow different dtypes and devices. See the link below. 

https://github.com/pytorch/pytorch/blob/095cd6a0da255a5ed24d146898c53790f0e9c3f0/torch/lib/c10d/reducer.cpp#L1897


However, addressing different dtypes is also out of the scope of this this task. Let's create an issue and add the URL here to track. We can come back to this if bandwidth allows.

Reply:
https://github.com/pytorch/pytorch/issues/59972